src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -71,6 +71,42 @@ public List<TElement> ToList()
             return list;
         }
 
+        public HashSet<TElement> ToHashSet(IEqualityComparer<TElement> comparer) => new HashSet<TElement>(_source, comparer);
+
+        public HashSet<TElement> ToHashSet(int minIdx, int maxIdx, IEqualityComparer<TElement> comparer)
+        {
+            Buffer<TElement> buffer = new Buffer<TElement>(_source);
+
+            int count = buffer._count;
+
+            if (count <= minIdx)
+            {
+                return new HashSet<TElement>(comparer);
+            }
+
+            if (count <= maxIdx)
+            {
+                maxIdx = count - 1;
+            }
+
+            if (minIdx == maxIdx)
+            {
+                return new HashSet<TElement>(1, comparer) { GetEnumerableSorter().ElementAt(buffer._items, count, minIdx) };
+            }
+
+            int[] map = SortedMap(buffer, minIdx, maxIdx);
```

Comment:
Incidentally, the `EnumerableSorter<TElement>.QuickSelect` method for finding a single element partitions around the element found as a byproduct. If it was changed to be `QuickSelect(int[] map, int left, int right, int idx)` rather than have `int left = 0` as the first line, then the current calls could be satisfied by `private int QuickSelect(int[] map, int right, int idx) => QuickSelect(map, 0, right, idx)`. We could then do:

```C#
int[] map = ComputeMap(elements, count);
QuickSelect(map, count - 1, minIdx);
QuickSelect(map, minIdx, count - 1, maxIdx);
```
And we should have a our partition between `minIdx` and `maxIdx` in O(1) time (indeed, in less than double the time the `ElementAt()` is on `OrderBy` results). This would have to be verified though; in particular I make no promises there isn't an off-by-one error in this top-of-my head comment code.